### ISSUE TITLE
Support ObjectCreationHandling Serialisation Option

### ISSUE THREAD

**Issue by flytzen** [Azure/azure-cosmos-dotnet-v3#2968]
# **Is your feature request related to a problem? Please describe.**
TL;DR; If a user assigns a *default value* to properties on objects they store in CosmosDB, then they won't get those properties back from the database when they retrieve the object (despite the data actually being in the database).

NewtonSoft.Json won't deserialise properties that have a default value, ie `public FieldClass SomeField { get; set; } = new FieldClass(0);`
See https://github.com/JamesNK/Newtonsoft.Json/issues/1675
The fix is to set the ObjectCreationHandling serialisation option to Replace - but this is not exposed in the SDK.

# **Describe the solution you'd like**
Expose the `ObjectCreationHandling` option in `CosmosSerializationOptions`.

# **Describe alternatives you've considered**
The user can work around this by using a custom serialiser (though this can be daunting to setup, especially in an existing codebase used in production).

# **Additional context**
If a user stores an object in a property that always has a default value then it won't be deserialised from the database.
Examples include 
`public Foo Foo { get => this.foo ?? new(); set => this.foo = value ?? new() }`
and
`public FieldClass SomeField { get; set; } = new FieldClass(0);`

This leads to hard-to-diagnose errors as they will only surface when objects are stored and retrieved and not in unit tests.
The problem is discussed here: https://github.com/JamesNK/Newtonsoft.Json/issues/1675

With NewtonSoft.Json you can fix this by [setting ObjectCreationHandling to Replace](https://www.newtonsoft.com/json/help/html/DeserializeObjectCreationHandling.htm)

*However, this option is not exposed in `CosmosSerializationOptions` meaning a user can't fix this with Cosmos without using a custom serialiser (which can be daunting to setup, especially in an existing codebase used in production).*

System.Text.Json [replaces by default](https://docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-migrate-from-newtonsoft-how-to?pivots=dotnet-6-0#reuse-rather-than-replace-properties) but at the moment the serialiser in Cosmos seems to be NewtonSoft. 

A downside to doing this would be if you have future plans to move to System.Text.Json as that does not have an equivalent option. However, making that move would change the default behaviour anyway and so would be a breaking change.

**Comment by ealsur** [comment]
You can customize the serializer completely, for example, you can use System.Text.Json https://github.com/Azure/azure-cosmos-dotnet-v3/tree/master/Microsoft.Azure.Cosmos.Samples/Usage/SystemTextJson or simply have one using Newtonsoft.Json (like the one the SDK uses https://github.com/Azure/azure-cosmos-dotnet-v3/blob/master/Microsoft.Azure.Cosmos/src/Serializer/CosmosJsonDotNetSerializer.cs) but with your own custom configuration.

**Comment by flytzen** [comment]
> You can customize the serializer completely....

@ealsur Thank you. Yes, I understand that :) However, that is a bigger thing to do and potentially quite daunting if your code is already in use in production: By adding a custom serialiser you run the risk of inadvertently changing away from the standard settings in the SDK and breaking something. You can, of course, do it and you can look in this repo and copy the entire class you mentioned and just make that one change. But, it's more work and more risk. 
Ideally the SDK should probably set this property Replace by default as that is the most logical for a database (not receiving back properties that exist in the database is not logical :)) - but I realise that would be a breaking change.

**Comment by ealsur** [comment]
Right, the point is, by guideline of the global Azure SDKs, the public surface cannot have anything specific to an external library (so we cannot expose anything that takes in the Newtonsoft types).
The custom serializer is only used to deserialize customer data, not anything internal. The "danger" of adding a custom serializer is only if you have any data types that will behave differently with the new serializer when converting the raw json.
You are right that changing this internally might be a breaking change if it impacts other users.

**Comment by flytzen** [comment]
Yeah, that makes sense. Not an easy thing :)
I guess the guidelines wouldn't preclude you from exposing your own version of "ObjectCreationHandling" - but it would tie your public interface uncomfortably close to an internal implementation detail. 
I don't know the right answer - but I can see this is something that will catch people out and can definitely lead to actual data loss if they save the retrieved object back to the database.
